firebase-database/src/main/java/com/google/firebase/database/core/Repo.java

```diff
@@ -522,36 +527,33 @@ public void run() {
               return;
             }
             serverSyncTree.setQueryActive(query.getSpec());
+            final DataSnapshot persisted = serverSyncTree.persistenceServerCache(query);
+            if (persisted.exists()) {
+              // Prefer the locally persisted value if the server is not responsive.
+              scheduleDelayed(() -> source.trySetResult(persisted), 3000);
+            }
             connection
                 .get(query.getPath().asList(), query.getSpec().getParams().getWireProtocolParams())
                 .addOnCompleteListener(
                     ((DefaultRunLoop) ctx.getRunLoop()).getExecutorService(),
-                    new OnCompleteListener<Object>() {
-                      @Override
-                      public void onComplete(@NonNull Task<Object> task) {
-                        if (!task.isSuccessful()) {
-                          operationLogger.info(
-                              "get for query "
-                                  + query.getPath()
-                                  + " falling back to disk cache after error: "
-                                  + task.getException().getMessage());
-                          DataSnapshot cached = serverSyncTree.persistenceServerCache(query);
-                          if (!cached.exists()) {
-                            source.setException(task.getException());
-                          } else {
-                            source.setResult(cached);
-                          }
-                        } else {
-                          Node serverNode = NodeUtilities.NodeFromJSON(task.getResult());
-                          postEvents(
-                              serverSyncTree.applyServerOverwrite(query.getPath(), serverNode));
-                          source.setResult(
-                              InternalHelpers.createDataSnapshot(
-                                  query.getRef(),
-                                  IndexedNode.from(serverNode, query.getSpec().getIndex())));
+                    (@NonNull Task<Object> task) -> {
+                      if (source.getTask().isComplete()) {
+                        return;
+                      }
+                      if (!task.isSuccessful()) {
+                        if (!persisted.exists()) {
```

Comment:
If the task is not successful and we do have a persisted value we should set it right away. We should not wait for 3s.